cloud/aws/services/ec2/natgateway.go

```diff
@@ -0,0 +1,126 @@
+package ec2
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+)
+
+func (s *Service) reconcileNatGateways(subnets []*v1alpha1.Subnet, vpc *v1alpha1.VPC) error {
+	if !s.containsPrivateSubnets(subnets) {
+		return nil
+	}
+
+	existing, err := s.describeNatGateways(vpc.ID)
+	if err != nil {
+		return err
+	}
+
+	for _, sn := range subnets {
+		if sn.ID == "" || !sn.IsPublic {
+			continue
+		}
+		if _, ok := existing[sn.ID]; ok {
+			continue
+		}
+
+		_, err := s.createNatGateway(sn.ID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) describeNatGateways(vpcID string) (map[string]*ec2.NatGateway, error) {
+	describeNatGatewayInput := ec2.DescribeNatGatewaysInput{
+		Filter: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{aws.String(vpcID)},
+			},
+		},
+	}
+
+	gateways := make(map[string]*ec2.NatGateway)
+
+	err := s.ec2.DescribeNatGatewaysPages(
+		&describeNatGatewayInput,
+		func(page *ec2.DescribeNatGatewaysOutput, lastPage bool) bool {
+			for _, r := range page.NatGateways {
+				gateways[*r.SubnetId] = r
+			}
+			return false
+		})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to describe NAT gateways with VPC ID %q", vpcID)
+	}
+	return gateways, nil
+}
+
+func (s *Service) allocateAddress() (string, error) {
```

Comment:
Should this method be moved somewhere else? I'm not totally sure if the gateway file is the best place for it.

Reply:
It's been moved. Was there for the convenience of not opening a new file at the time.